### ISSUE TITLE
Matching tokens with &str field?

### ISSUE THREAD

**Issue by osa1** [lalrpop/lalrpop#556]
In my lexer I have this token type:

```
enum Token<'input> {
    Id(&'input str),
    ...
}
```

In my grammar I'm trying to do something like

```
use lexer::{Token, LexerError};

grammar<'input>(input: &'input str);

extern {
    type Location = usize;
    type Error = LexerError;

    enum Token {
        "foo" => Token::Id("foo"),
        "bar" => Token::Id("bar"),
        ...
    }
}
```

Normally I'd have different `Token` variants for these identifiers, but this lexer is used by multiple parsers, each have different set of "special" identifiers.

Is this currently possible? If not, would it be possible to extend lalrpop to do this? I'm happy to look into this is someone can advise.

**Comment by brendanzab** [comment]
I've wanted this in the past too!

Another similar issue is if you have an enum in the token, eg:

```
"{" => Token::Open(Delim::Brace),
"}" => Token::Close(Delim::Brace),
```

**Comment by osa1** [comment]
Yeah I think your example makes sense too.

Can someone who knows the code base comment on how difficult this would be to implement?

**Comment by Marwes** [comment]
Probably not too difficult, you will want to modify this part of the parser https://github.com/lalrpop/lalrpop/blob/149e77b301828392b09da5c2ba634fd19ce3a3a7/lalrpop/src/parser/lrgrammar.lalrpop#L356-L383 to accept these cases. Then follow the compilation errors to fill in the rest (I think it mostly involves updating enums to accept more and extend the code generation to write the right things).

**Comment by osa1** [comment]
Thanks @Marwes , I will give it a try.

**Comment by osa1** [comment]
@brendanzab it seems like your case is already supported, at least in `master`. Here's an example grammar:

```
grammar;

use util::tok::{Tok, Delim};

extern {
    enum Tok {
        "{" => Tok::Open(Delim::Brace),
        "}" => Tok::Close(Delim::Brace),
        "[" => Tok::Open(Delim::Bracket),
        "]" => Tok::Close(Delim::Bracket),
    }
}

pub E: i32 = {
    => 0,
    "{" <e:E> => e + 1,
    "}" <e:E> => e - 1,
    "[" <e:E> => e + 1,
    "]" <e:E> => e - 1,
};
```

`tok.rs`:

<details>

<summary>tok.rs</summary>

```rust
use std::str::FromStr;

#[derive(Debug, PartialEq, Eq)]
pub struct NoCloneTok(pub Tok);

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Tok {
    Num(i32),
    LParen,
    RParen,
    Minus,
    Plus,
    Times,
    Div,
    Comma,
    Open(Delim),
    Close(Delim),
    #[allow(dead_code)]
    Fraction(i32, i32), // Not produced by tokenizer, used only in regression tests for #179
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Delim {
    /// '{' or '}'
    Brace,
    /// '[' or ']'
    Bracket,
}

// simplest and stupidest possible tokenizer
pub fn tokenize(s: &str) -> Vec<(usize, Tok, usize)> {
    let mut tokens = vec![];
    let mut chars = s.chars();
    let mut lookahead = chars.next();
    while let Some(c) = lookahead {
        // skip whitespace characters
        if !c.is_whitespace() {
            match c {
                '(' => tokens.push(Tok::LParen),
                ')' => tokens.push(Tok::RParen),
                '-' => tokens.push(Tok::Minus),
                '+' => tokens.push(Tok::Plus),
                '*' => tokens.push(Tok::Times),
                ',' => tokens.push(Tok::Comma),
                '/' => tokens.push(Tok::Div),
                '{' => tokens.push(Tok::Open(Delim::Brace)),
                '[' => tokens.push(Tok::Open(Delim::Bracket)),
                '}' => tokens.push(Tok::Close(Delim::Brace)),
                ']' => tokens.push(Tok::Close(Delim::Bracket)),
                _ if c.is_digit(10) => {
                    let (tmp, next) = take_while(c, &mut chars, |c| c.is_digit(10));
                    lookahead = next;
                    tokens.push(Tok::Num(i32::from_str(&tmp).unwrap()));
                    continue;
                }
                _ => {
                    panic!("invalid character: {:?}", c);
                }
            }
        }

        // advance to next character by default
        lookahead = chars.next();
    }

    tokens
        .into_iter()
        .enumerate()
        .map(|(i, tok)| (i * 2, tok, i * 2 + 1))
        .collect()
}

fn take_while<C, F>(c0: char, chars: &mut C, f: F) -> (String, Option<char>)
where
    C: Iterator<Item = char>,
    F: Fn(char) -> bool,
{
    let mut buf = String::new();

    buf.push(c0);

    for c in chars {
        if !f(c) {
            return (buf, Some(c));
        }

        buf.push(c);
    }

    (buf, None)
}
```

</details>

This works just fine. It seems like we just need to accept string literal syntax in patterns for my use case and it should just work.

**Comment by osa1** [comment]
I implemented this in #557. I also added some tests similar to @brendanzab 's example above.
